File: packages/compress-stringify/python/itkwasm-compress-stringify/itkwasm_compress_stringify/interface_type_json.py
from dataclasses import asdict
import json

from itkwasm import Image, FloatTypes, Mesh, PolyData, array_like_to_bytes, buffer_to_numpy_array, JsonCompatible

from .compress_stringify import compress_stringify
from .parse_string_decompress import parse_string_decompress

def image_to_json(image: Image) -> JsonCompatible:
    """Convert an image to a JSON compatible dict

    :param image: Input image
    :type  image: itkwasm.Image

    :return: Image dict with binary data compressed and converted to a string
    :rtype:  JsonCompatible
    """
    image_dict = asdict(image)
    level = 5

    direction_bytes = array_like_to_bytes(image_dict["direction"])
    image_dict["direction"] = compress_stringify(direction_bytes, compression_level=level, stringify=True).decode()

    if image_dict["data"] is not None:
        data_bytes = array_like_to_bytes(image_dict["data"])
        image_dict["data"] = compress_stringify(data_bytes, compression_level=level, stringify=True).decode()

    return image_dict

def json_to_image(image_json: JsonCompatible) -> Image:
    """Convert a JSON compatible image dict to an image

    :param image_json: Input image dict with binary data compressed and converted to a string
    :type  image_json: JsonCompatible

    :return: Output image
    :rtype:  itkwasm.Image
    """
    image_dict = image_json
    image_type = image_dict["imageType"]
    dimension = image_type["dimension"]

    direction_bytes = parse_string_decompress(image_dict["direction"].encode(), parse_string=True)
    image_dict["direction"] = buffer_to_numpy_array(FloatTypes.Float64, direction_bytes).reshape((dimension, dimension))

    if image_dict["data"] is not None:
        data_bytes = parse_string_decompress(image_dict["data"].encode(), parse_string=True)
        data_array = buffer_to_numpy_array(image_type["componentType"], data_bytes)
        shape = list(image_dict["size"])[::-1]
        if image_type["components"] > 1:
            shape.append(image_type["components"])
        image_dict["data"] = data_array.reshape(tuple(shape))

    image = Image(**image_dict)

    return image

def mesh_to_json(mesh: Mesh) -> JsonCompatible:
    """Convert an mesh to a JSON compatible dict

    :param mesh: Input mesh
    :type  mesh: itkwasm.Mesh

    :return: Mesh dict with binary data compressed and converted to a string
    :rtype:  JsonCompatible
    """
    mesh_dict = asdict(mesh)
    level = 5

    for key in ["points", "pointData", "cells", "cellData"]:
        if mesh_dict[key] is not None:
            mesh_dict[key] = compress_stringify(array_like_to_bytes(mesh_dict[key]), compression_level=level, stringify=True).decode()

    return mesh_dict

def json_to_mesh(mesh_json: JsonCompatible) -> Mesh:
    """Convert a JSON compatible mesh dict to an mesh

    :param mesh_json: Input mesh dict with binary data compressed and converted to a string
    :type  mesh_json: JsonCompatible

    :return: Output mesh
    :rtype:  itkwasm.Mesh
    """
    mesh_dict = mesh_json

    for key in ["points", "pointData", "cells", "cellData"]:
        if mesh_dict[key] is not None:
            mesh_dict[key] = parse_string_decompress(mesh_dict[key].encode(), parse_string=True)

    mesh = Mesh(**mesh_dict)
    return mesh

def poly_data_to_json(poly_data: PolyData) -> JsonCompatible:
    """Convert an PolyData to a JSON compatible dict

    :param poly_data: Input PolyData
    :type  poly_data: itkwasm.PolyData

    :return: PolyData dict with binary data compressed and converted to a string
    :rtype:  JsonCompatible
    """
    poly_data_dict = asdict(poly_data)
    level = 5

    for key in ["points", "vertices", "lines", "polygons", "triangleStrips", "pointData", "cellData"]:
        if poly_data_dict[key] is not None:
            poly_data_dict[key] = compress_stringify(array_like_to_bytes(poly_data_dict[key]), compression_level=level, stringify=True).decode()

    return poly_data_dict

def json_to_poly_data(poly_data_json: str) -> PolyData:
    """Convert a JSON compatible PolyData dict to a PolyData

    :param poly_data_json: Input PolyData dict with binary data compressed and converted to a string
    :type  poly_data_json: JsonCompatible

    :return: Output poly_data
    :rtype:  itkwasm.PolyData
    """
    poly_data_dict = poly_data_json

    for key in ["points", "vertices", "lines", "polygons", "triangleStrips", "pointData", "cellData"]:
        if poly_data_dict[key] is not None:
            poly_data_dict[key] = parse_string_decompress(poly_data_dict[key].encode(), parse_string=True)

    poly_data = PolyData(**poly_data_dict)
    return poly_data

File: packages/compress-stringify/typescript/src/interface-type-json.ts
import {
  bufferToTypedArray,
  WorkerPoolFunctionOption,
  WorkerPoolFunctionResult,
  Image,
  Mesh,
  PolyData,
} from "itk-wasm";

import compressStringify from "./compress-stringify.js";
import parseStringDecompress from "./parse-string-decompress.js";

import {
  ImageJson,
  MeshJson,
  PolyDataJson,
} from "./interface-type-json-common.js";

export interface ImageToJsonResult extends WorkerPoolFunctionResult {
  encoded: ImageJson;
}

export interface ImageToJsonOptions extends WorkerPoolFunctionOption {}

export async function imageToJson(
  image: Image,
  options: ImageToJsonOptions = {},
): Promise<ImageToJsonResult> {
  const level = 5;

  const encoded = new Image(image.imageType) as unknown as ImageJson;
  encoded.origin = image.origin;
  encoded.spacing = image.spacing;
  encoded.size = image.size;
  encoded.metadata = image.metadata;

  const decoder = new TextDecoder("utf-8");
  const directionBytes = new Uint8Array(image.direction.buffer);
  const direction = await compressStringify(directionBytes, {
    compressionLevel: level,
    stringify: true,
    webWorker: options.webWorker,
    noCopy: options.noCopy,
  });
  const usedWebWorker = direction.webWorker;
  encoded.direction = decoder.decode(direction.output.buffer);

  if (image.data === null) {
    encoded.data = null;
  } else {
    const dataBytes = new Uint8Array(image.data.buffer);
    const encodedData = await compressStringify(dataBytes, {
      compressionLevel: level,
      stringify: true,
      webWorker: usedWebWorker,
      noCopy: options.noCopy,
    });
    encoded.data = decoder.decode(encodedData.output.buffer);
  }

  return { encoded, webWorker: usedWebWorker };
}

export interface JsonToImageResult extends WorkerPoolFunctionResult {
  decoded: Image;
}

export interface JsonToImageOptions extends WorkerPoolFunctionOption {}

export async function jsonToImage(
  encoded: ImageJson,
  options: JsonToImageOptions = {},
): Promise<JsonToImageResult> {
  const decoded = encoded as unknown as Image;

  const encoder = new TextEncoder();
  const directionBytes = new Uint8Array(encoder.encode(encoded.direction));
  const direction = await parseStringDecompress(directionBytes, {
    parseString: true,
    webWorker: options.webWorker,
    noCopy: true,
  });
  const usedWebWorker = direction.webWorker;
  decoded.direction = new Float64Array(direction.output.buffer);

  if (!decoded.data) {
    decoded.data = null;
  } else {
    const dataBytes = new Uint8Array(encoder.encode(encoded.data as string));
    const decodedData = await parseStringDecompress(dataBytes, {
      parseString: true,
      webWorker: usedWebWorker,
      noCopy: true,
    });
    decoded.data = bufferToTypedArray(
      decoded.imageType.componentType,
      decodedData.output.buffer,
    );
  }

  return { decoded, webWorker: usedWebWorker };
}

export interface MeshToJsonResult extends WorkerPoolFunctionResult {
  encoded: MeshJson;
}

export interface MeshToJsonOptions extends WorkerPoolFunctionOption {}

export async function meshToJson(
  mesh: Mesh,
  options: MeshToJsonOptions = {},
): Promise<MeshToJsonResult> {
  const level = 5;

  const encoded = new Mesh(mesh.meshType) as unknown as MeshJson;
  encoded.name = mesh.name;
  encoded.numberOfPoints = mesh.numberOfPoints;
  encoded.numberOfCells = mesh.numberOfCells;
  encoded.numberOfPointPixels = mesh.numberOfPointPixels;
  encoded.numberOfCellPixels = mesh.numberOfCellPixels;
  encoded.cellBufferSize = mesh.cellBufferSize;

  let usedWebWorker = options.webWorker;
  const decoder = new TextDecoder("utf-8");
  for (const prop of ["points", "pointData", "cells", "cellData"]) {
    // @ts-ignore: TS7053
    if (mesh[prop] === null) {
      // @ts-ignore: TS7053
      encoded[prop] = null;
    } else {
      // @ts-ignore: TS7053
      const dataBytes = new Uint8Array(mesh[prop].buffer);
      const encodedProp = await compressStringify(dataBytes, {
        compressionLevel: level,
        stringify: true,
        webWorker: usedWebWorker,
        noCopy: options.noCopy,
      });
      usedWebWorker = encodedProp.webWorker;
      // @ts-ignore
      encoded[prop] = decoder.decode(encodedProp.output.buffer);
    }
  }

  return { encoded, webWorker: usedWebWorker as Worker };
}

export interface JsonToMeshResult extends WorkerPoolFunctionResult {
  decoded: Mesh;
}

export interface JsonToMeshOptions extends WorkerPoolFunctionOption {}

export async function jsonToMesh(
  encoded: MeshJson,
  options: JsonToMeshOptions = {},
): Promise<JsonToMeshResult> {
  // @ts-ignore: TS7053
  const decoded = encoded as Mesh;

  const componentTypeMap = new Map([
    ["points", "pointComponentType"],
    ["pointData", "pointPixelComponentType"],
    ["cells", "cellComponentType"],
    ["cellData", "cellPixelComponentType"],
  ]);

  const encoder = new TextEncoder();
  let usedWebWorker = options.webWorker;
  for (const prop of ["points", "pointData", "cells", "cellData"]) {
    // @ts-ignore: TS7053
    if (decoded[prop] === null) {
      // @ts-ignore: TS7053
      decoded[prop] = null;
    } else {
      // @ts-ignore: TS7053
      const dataBytes = new Uint8Array(encoder.encode(decoded[prop]));
      const decodedProp = await parseStringDecompress(dataBytes, {
        parseString: true,
        webWorker: usedWebWorker,
        noCopy: true,
      });
      usedWebWorker = decodedProp.webWorker;
      // @ts-ignore
      decoded[prop] = bufferToTypedArray(
        // @ts-ignore: TS7053
        decoded.meshType[componentTypeMap.get(prop)],
        decodedProp.output.buffer,
      );
    }
  }

  return { decoded, webWorker: usedWebWorker as Worker };
}

export interface PolyDataToJsonResult extends WorkerPoolFunctionResult {
  encoded: PolyDataJson;
}

export interface PolyDataToJsonOptions extends WorkerPoolFunctionOption {}

export async function polyDataToJson(
  polyData: PolyData,
  options: PolyDataToJsonOptions = {},
): Promise<PolyDataToJsonResult> {
  const level = 5;

  const encoded = new PolyData(
    polyData.polyDataType,
  ) as unknown as PolyDataJson;
  encoded.name = polyData.name;
  encoded.numberOfPoints = polyData.numberOfPoints;
  encoded.verticesBufferSize = polyData.verticesBufferSize;
  encoded.linesBufferSize = polyData.linesBufferSize;
  encoded.polygonsBufferSize = polyData.polygonsBufferSize;
  encoded.triangleStripsBufferSize = polyData.triangleStripsBufferSize;
  encoded.numberOfPointPixels = polyData.numberOfPointPixels;
  encoded.numberOfCellPixels = polyData.numberOfCellPixels;

  let usedWebWorker = options.webWorker;
  const decoder = new TextDecoder("utf-8");
  for (const prop of [
    "points",
    "vertices",
    "lines",
    "polygons",
    "triangleStrips",
    "pointData",
    "cellData",
  ]) {
    // @ts-ignore: TS7053
    if (polyData[prop] === null) {
      // @ts-ignore: TS7053
      encoded[prop] = null;
    } else {
      // @ts-ignore: TS7053
      const dataBytes = new Uint8Array(polyData[prop].buffer);
      // @ts-ignore
      const encodedProp = await compressStringify(dataBytes, {
        compressionLevel: level,
        stringify: true,
        webWorker: usedWebWorker,
        noCopy: options.noCopy,
      });
      usedWebWorker = encodedProp.webWorker;
      // @ts-ignore
      encoded[prop] = decoder.decode(encodedProp.output.buffer);
    }
  }

  return { encoded, webWorker: usedWebWorker as Worker };
}

export interface JsonToPolyDataResult extends WorkerPoolFunctionResult {
  decoded: PolyData;
}

export interface JsonToPolyDataOptions extends WorkerPoolFunctionOption {}

export async function jsonToPolyData(
  encoded: PolyDataJson,
  options: JsonToPolyDataOptions = {},
): Promise<JsonToPolyDataResult> {
  // @ts-ignore: TS7053
  const decoded = encoded as PolyData;

  const componentTypeMap = new Map([
    ["points", "float32"],
    ["vertices", "uint32"],
    ["lines", "uint32"],
    ["polygons", "uint32"],
    ["triangleStrips", "uint32"],
    ["pointData", decoded.polyDataType.pointPixelComponentType],
    ["cellData", decoded.polyDataType.cellPixelComponentType],
  ]);

  const encoder = new TextEncoder();
  let usedWebWorker = options.webWorker;
  for (const prop of [
    "points",
    "vertices",
    "lines",
    "polygons",
    "triangleStrips",
    "pointData",
    "cellData",
  ]) {
    // @ts-ignore: TS7053
    if (decoded[prop] === null) {
      // @ts-ignore: TS7053
      decoded[prop] = null;
    } else {
      // @ts-ignore: TS7053
      const dataBytes = new Uint8Array(encoder.encode(decoded[prop]));
      // @ts-ignore
      const decodedProp = await parseStringDecompress(dataBytes, {
        parseString: true,
        webWorker: usedWebWorker,
        noCopy: true,
      });
      usedWebWorker = decodedProp.webWorker;
      // @ts-ignore
      decoded[prop] = bufferToTypedArray(
        // @ts-ignore: TS7053
        componentTypeMap.get(prop),
        decodedProp.output.buffer,
      );
    }
  }

  return { decoded, webWorker: usedWebWorker as Worker };
}
